### PR TITLE
Fix 'reinterpret' casts of StringExp to static array with implicit extension

### PR DESCRIPTION
**Pull request by kinke**
Fixes issue 19639.

### CODE REVIEW

File: test/runnable/test19639.d [hunk 1]
```diff
@@ -0,0 +1,22 @@
+enum EMPTY_STRING = ""[0..0];
+enum SMALL_STRING = "a"[0..1];
+
+void main()
+{
+    {
+        char[64] buf = EMPTY_STRING;
```
Comment (kinke) [inline test/runnable/test19639.d]: Currently segfaults (the `EMPTY_STRING` buffer (`StringExp.string`) is null).

File: test/runnable/test19639.d [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        foreach (c; buf) assert(c == 0);
+        buf[$-1] = 'e';
+        buf = EMPTY_STRING;
+        assert(buf[$-1] == 0);
+    }
+
+    {
+        char[64] buf = SMALL_STRING;
+        assert(buf[0] == 'a');
+        foreach (c; buf[1..$]) assert(c == 0);
```
Comment (kinke) [inline test/runnable/test19639.d]: Currently likely to fail and/or crash as the compile-time memcpy reads beyond the `SMALL_STRING` buffer.

**Comment by Basile-z** [comment]
This looks good but it PR can be submitted for the stable branch as it fixes a regression.

**Comment by kinke** [comment]
I had it on stable and then switched to master, as I deemed it better suited (sooner in nightlies). It's a sort of regression - the crash for the first test block happens since v2.071, while the read-beyond-src-buffer was apparently there since v2.068 (https://run.dlang.io/is/oPq799).

Implicit extension was apparently added with v2.066.

**Review by ghost — APPROVED** [review]
all right, pink @thewilsonator

**Review by thewilsonator — APPROVED** [review]

**Comment by kinke** [comment]
Hmm, the bugzilla issue wasn't auto-closed...

**Comment by wilzbach** [comment]
> Hmm, the bugzilla issue wasn't auto-closed.

The Bugzilla integration closes only things when they are merged in master automatically. Stable was turned off as too many people complained about lots of emails on new releases. Though IMHO that's a lot less worse than an integration which doesn't seem to work...

**Comment by kinke** [comment]
@wilzbach: That's why I was wondering, this targeted master.

**Comment by wilzbach** [comment]
Sorry. Only read this from my phone. Then I also don't know what went wrong :man_shrugging: 
Maybe because GitHub services are deprecated and to be removed eventually?

image

**Comment by wilzbach** [comment]
Ah well, there's an explanation: https://developer.github.com/changes/2018-04-25-github-services-deprecation

---
May 31, 2018: Submit your intentions for migrating or replacing your GitHub Service.
October 1, 2018: GitHub discontinues allowing users to install services. GitHub Services will be removed from the GitHub.com UI.
January 29, 2019: As an alternative to the email service, you can now start using email notifications for pushes to your repository. See "About email notifications for pushes to your repository" in the GitHub Help documentation to learn how to configure commit email notifications.
January 31, 2019: GitHub stops delivering installed services' events on GitHub.com.
---

**Pull request merged**